File: VB/WebSite/App_Code/CustomControllerAndCommands.vb
Imports System
Imports DevExpress.XtraScheduler
Imports DevExpress.Web.ASPxScheduler
Imports DevExpress.Web.ASPxScheduler.Internal

Public Class CustomInsertAppointmentCommand
    Inherits AppointmentClientSideInsertCallbackCommand

    Public Sub New(ByVal scheduler As ASPxScheduler)
        MyBase.New(scheduler)
    End Sub

    Protected Overrides Function CreateAppointmentFormController(ByVal apt As Appointment) As AppointmentFormController
        Return New CustomAppointmentFormController(Control, apt)
    End Function

    Protected Overrides Sub AssignControllerCustomFieldsValues(ByVal controller As AppointmentFormController, ByVal clientAppointment As ClientAppointmentProperties)
        Dim editedAppointment As Appointment = controller.EditedAppointmentCopy

        If clientAppointment.Properties("Price") IsNot Nothing Then
            Dim result As Double = 0
            Double.TryParse(clientAppointment.Properties("Price").ToString(), result)
            editedAppointment.CustomFields("Price") = result
        End If

        MyBase.AssignControllerCustomFieldsValues(controller, clientAppointment)
    End Sub
End Class

Public Class CustomUpdateAppointmentCommand
    Inherits AppointmentClientSideUpdateCallbackCommand

    Public Sub New(ByVal scheduler As ASPxScheduler)
        MyBase.New(scheduler)
    End Sub

    Protected Overrides Function CreateAppointmentFormController(ByVal apt As Appointment) As AppointmentFormController
        Return New CustomAppointmentFormController(Control, apt)
    End Function

    Protected Overrides Sub AssignControllerCustomFieldsValues(ByVal controller As AppointmentFormController, ByVal clientAppointment As ClientAppointmentProperties)
        Dim editedAppointment As Appointment = controller.EditedAppointmentCopy

        If clientAppointment.Properties("Price") IsNot Nothing Then
            Dim result As Double = 0
            Double.TryParse(clientAppointment.Properties("Price").ToString(), result)
            editedAppointment.CustomFields("Price") = result
        Else
            editedAppointment.CustomFields("Price") = Nothing
        End If

        MyBase.AssignControllerCustomFieldsValues(controller, clientAppointment)
    End Sub
End Class

Public Class CustomAppointmentFormController
    Inherits AppointmentFormController

    Private Const PriceFieldName As String = "Price"

    Public Sub New(ByVal control As ASPxScheduler, ByVal apt As Appointment)
        MyBase.New(control, apt)
    End Sub

    Public Property Price() As Double?
        Get
            Return CType(EditedAppointmentCopy.CustomFields(PriceFieldName), Double?)
        End Get
        Set(ByVal value? As Double)
            EditedAppointmentCopy.CustomFields(PriceFieldName) = value
        End Set
    End Property

    Private Property SourcePrice() As Double?
        Get
            Return CType(SourceAppointment.CustomFields(PriceFieldName), Double?)
        End Get
        Set(ByVal value? As Double)
            SourceAppointment.CustomFields(PriceFieldName) = value
        End Set
    End Property

    Protected Overrides Sub ApplyCustomFieldsValues()
        SourcePrice = Price
    End Sub
End Class

File: CS/WebSite/App_Code/CustomControllerAndCommands.cs
using System;
using DevExpress.XtraScheduler;
using DevExpress.Web.ASPxScheduler;
using DevExpress.Web.ASPxScheduler.Internal;

public class CustomInsertAppointmentCommand : AppointmentClientSideInsertCallbackCommand {
    public CustomInsertAppointmentCommand(ASPxScheduler scheduler)
        : base(scheduler) {
    }

    protected override AppointmentFormController CreateAppointmentFormController(Appointment apt) {
        return new CustomAppointmentFormController(Control, apt);
    }

    protected override void AssignControllerCustomFieldsValues(AppointmentFormController controller, ClientAppointmentProperties clientAppointment) {
        Appointment editedAppointment = controller.EditedAppointmentCopy;

        if (clientAppointment.Properties["Price"] != null) {
            double result = 0;
            Double.TryParse(clientAppointment.Properties["Price"].ToString(), out result);
            editedAppointment.CustomFields["Price"] = result;
        }

        base.AssignControllerCustomFieldsValues(controller, clientAppointment);
    }
}

public class CustomUpdateAppointmentCommand : AppointmentClientSideUpdateCallbackCommand {
    public CustomUpdateAppointmentCommand(ASPxScheduler scheduler)
        : base(scheduler) {
    }

    protected override AppointmentFormController CreateAppointmentFormController(Appointment apt) {
        return new CustomAppointmentFormController(Control, apt);
    }

    protected override void AssignControllerCustomFieldsValues(AppointmentFormController controller, ClientAppointmentProperties clientAppointment) {
        Appointment editedAppointment = controller.EditedAppointmentCopy;

        if (clientAppointment.Properties["Price"] != null) {
            double result = 0;
            Double.TryParse(clientAppointment.Properties["Price"].ToString(), out result);
            editedAppointment.CustomFields["Price"] = result;
        }
        else
            editedAppointment.CustomFields["Price"] = null;

        base.AssignControllerCustomFieldsValues(controller, clientAppointment);
    }
}

public class CustomAppointmentFormController : AppointmentFormController {
    private const string PriceFieldName = "Price";

    public CustomAppointmentFormController(ASPxScheduler control, Appointment apt)
        : base(control, apt) {
    }

    public double? Price {
        get { return (double?)EditedAppointmentCopy.CustomFields[PriceFieldName]; }
        set { EditedAppointmentCopy.CustomFields[PriceFieldName] = value; }
    }

    double? SourcePrice {
        get { return (double?)SourceAppointment.CustomFields[PriceFieldName]; }
        set { SourceAppointment.CustomFields[PriceFieldName] = value; }
    }

    protected override void ApplyCustomFieldsValues() {
        SourcePrice = Price;
    }
}